### PR TITLE
Add scikit-learn wrapper

### PR DESCRIPTION
**Pull request by shiodat**
This PR enables us to cooperate scikit-learn more easily.

We now can use jubatus with scikit-learn APIs like [Pipeline API](http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html) , [Cross Validation API](http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.cross_val_score.html#sklearn.cross_validation.cross_val_score), [Grid Search API](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV) and so on.

### CODE REVIEW

File: jubakit/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    self.fitted_ = True
+    return self
+
+  def fit(self, X, y):
+    """
+    Fit model.
+    """
+    self._launch_classifier()
+    if not self.embedded:
+      self.clf_.clear()
```
Comment (kmaehashi) [inline jubakit/wrapper.py]: Is these 3 lines needed? (just calling `partial_fit` seems fine)

File: jubakit/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    """
+    Creates a base class for Jubatus Classifiers"
+    """
+    self.softmax = softmax
+    self.n_iter = n_iter
+    self.shuffle = shuffle
+    self.embedded = embedded
+    self.seed = seed
+    self.fitted_ = False
+    self.clf_ = None
```
Comment (shiodat) [inline jubakit/wrapper.py]: Okay, I change `clf` to `classifier`
Note: scikit-learn developers use `clf` for classifier object.
https://github.com/scikit-learn/scikit-learn/blob/b6941847d6f224446917ba52b4467cfbf082fcd8/sklearn/ensemble/voting_classifier.py#L203

File: jubakit/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  def __init__(self, n_iter=1, shuffle=False, softmax=False, embedded=True, seed=None):
+    """
+    Creates a base class for Jubatus Classifiers"
+    """
+    self.softmax = softmax
+    self.n_iter = n_iter
+    self.shuffle = shuffle
+    self.embedded = embedded
+    self.seed = seed
+    self.fitted_ = False
```
Comment (shiodat) [inline jubakit/wrapper.py]: Scikit-learn APIs follow the following naming scheme.
- argument parameters are expressed without underscore, i.e. `self.shuffled`.
- parameters created during executions are expressed with underscore, i.e. `self.classes_`.
http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html#sklearn.linear_model.LogisticRegression

Which do you think we have to obey the scikit-learn APIs naming schema or not?

Actually, I notice that we don't have to use `self.fitted_` parameter because if classifier is not trained, `self.classifier_` object becomes None, and I remove `self.fitted_` parameter.

File: jubakit/wrapper.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
+# -*- coding: utf-8 -*-
```
Comment (shiodat) [inline jubakit/wrapper.py]: Nice, I moved the file.

File: jubakit/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    Launch Jubatus Classifier 
+    """
+    raise NotImplementedError()
+
+  def partial_fit(self, X, y):
+    """
+    Partially fit underlying model.
+    If underlying model does not exist, launch a new model.
+    """
+    if getattr(self, 'classes_', None) is None:
```
Comment (shiodat) [inline jubakit/wrapper.py]: From scikit-learn Contribution pages
```
Attributes that have been estimated from the data must always have a name ending with trailing underscore, for example the coefficients of some regression estimator would be stored in a coef_ attribute after fit has been called.
```
The `classes_` parameter are estimated from the data, and I want to obey their coding style.

File: jubakit/wrapper.py (hunk 3 above)
Comment (kmaehashi) [inline jubakit/wrapper.py]: I see. As you proposed, using sklearn-style coding rule makes the code easier to understand for users with experience of sklearn.

File: jubakit/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+
+class BaseJubatusClassifier(BaseEstimator, ClassifierMixin):
+  """
+  scikit-learn Wrapper for Jubatus Classifiers.
+  """
+  
+  def __init__(self, n_iter=1, shuffle=False, softmax=False, embedded=True, seed=None):
+    """
+    Creates a base class for Jubatus Classifiers"
```
Comment (shiodat) [inline jubakit/wrapper.py]: Thanks, I fix it.

File: jubakit/wrapper.py (hunk 1 above)
Comment (shiodat) [inline jubakit/wrapper.py]: - `fit` trains `X` using new classifier model.
- `partial_fit` trains `X` with existing classifier model if exists.

If these 3 lines are removed, we cannot reuse classifier instance.
For example, grid search task reuses classifier instance, but if these 3 lines are removed, classifier with new method will not launched and same cv_score will be returned.
```
clf = LinearClassifier(**args)
for method in ['AROW', 'CW']:
   clf.set_params(method=method)
   clf.fit(X_train, y_train)
   y_pred = clf.predict(X_test)
   cv_score = accuracy(y_test, y_pred)
```

File: jubakit/wrapper.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    """
+    Partially fit underlying model.
+    If underlying model does not exist, launch a new model.
+    """
+    if getattr(self, 'classes_', None) is None:
+      self.classes_ = np.unique(y)
+    if self.clf_ is None:
+      self._launch_classifier()
+      if not self.embedded:
+        self.clf_.clear()
```
Comment (shiodat) [inline jubakit/wrapper.py]: Thank you for your comment, I notice that we don't have to exclude embedded case.

I intended that 
- default jubatus reuse the classifier model at assined port and host but  embedded jubatus always create new classifier model. 
- so classifier object is embedded, we don't have to clear model explicitly.

**Comment by shiodat** [comment]
@kmaehashi Thank you for review! 
I fixed as follows:
- Move `jubakit/wrapper,py` to `jubakit/wrapper/classifier.py`.
- Rename unclear attributes, i.e. `clf_` to `classifier_` and `cfg_` to `config_`.
- Remove unneeded lines if classifier type is embedded jubatus.

File: jubakit/wrapper.py (hunk 1 above)
Comment (kmaehashi) [inline jubakit/wrapper.py]: Ok, I understood what is intended here.

File: example/classifier_sklearn_grid_search.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+Grid Search with scikit-learn
+=============================
+
+In this example, we show how to use scikit-learn wrapper.
+
+"""
+
+import sklearn
+from operator import itemgetter
+from jubakit.wrapper import LinearClassifier, NearestNeighborsClassifier
```
Comment (kmaehashi) [inline example/classifier_sklearn_grid_search.py]: `jubakit.wrapper.classifier`

File: example/classifier_sklearn_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+""" 
+Pipeline Classification with scikit-learn wrapped Classifier
+============================================================
+
+In this example, we show how to use scikit-learn wrapper's 
+`fit(X, y)` and `predict(X)` functions. 
+
+"""
+
+from jubakit.wrapper import LinearClassifier, NearestNeighborsClassifier
```
Comment (kmaehashi) [inline example/classifier_sklearn_wrapper.py]: `jubakit.wrapper.classifier`

File: jubakit/wrapper/classifier.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+from .classifier import Classifier, Config, Dataset
```
Comment (kmaehashi) [inline jubakit/wrapper/classifier.py]: `.classifier` -> `..classifier`

File: jubakit/wrapper/classifier.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
+# -*- coding: utf-8 -*-
```
Comment (kmaehashi) [inline jubakit/wrapper/classifier.py]: You need to place `jubakit/wrapper/__init__.py` (otherwise it won't work)

**Comment by kmaehashi** [comment]
👍 Thank you!
